### PR TITLE
Admission response

### PR DESCRIPTION
**Pull request by cheftako**
**What this PR does / why we need it**: This splits the AdmissionReview into AdmissionRequest and AdmissionResponse. The AdmissionResponse the allows the mutating webhook to send back a json path with the mutated version of the requested object.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/features/issues/492

**Special notes for your reviewer**:Port of caesarxuchao's fix.

**Release note**:
```release-note
action-required: please update your admission webhook to use the latest [Admission API](https://github.com/kubernetes/api/tree/master/admission).

`admission/v1alpha1#AdmissionReview` is split into `AdmissionRequest` and `AdmissionResponse`. `AdmissionResponse` includes a `Patch` field to allow mutating webhooks to send json patch to the apiserver.
```

### CODE REVIEW

**Comment by liggitt** [comment]
Where was the discussion of returning a different top level type than posted?

File: pkg/apis/admission/types.go
Comment (liggitt) [inline pkg/apis/admission/types.go]: Why is this outside the status?

Comment (liggitt) [reply pkg/apis/admission/types.go]: Isn't the patch type required to be indicated?

**Comment by caesarxuchao** [comment]
> Where was the discussion of returning a different top level type than posted?

I thought the `AdmissionResponse` API was approved. The [design doc](https://docs.google.com/document/d/1BT8mZaT42jVxtC6l14YMXpUq0vZc6V5MPf_jnzDMMcg/edit#heading=h.j2g6525on607) says "The only approved section of this design doc is AdmissionResponse type described in the next sub-section."

Anyway, what's the benefit of using the same top-level type for both request and response?

> Why is this outside the status?

I guess the intention is to emphasis that only mutating webhooks can write `PatchBody`, but both mutating and validating webhooks can write `Status`.

Alternatively, how about flattening the "AdmissionResponse"? Like:

```golang
type AdmissionResponse struct {
  metav1.TypeMeta
  Allowed bool
  Result *metav1.Status
  PatchBody []byte
}
```

> Isn't the patch type required to be indicated?

I think the decision was to only support json-patch to begin with.

**Comment by liggitt** [comment]
> I thought the AdmissionResponse API was approved. The design doc says "The only approved section of this design doc is AdmissionResponse type described in the next sub-section."

I'm not seeing that comment, and the type changes aren't in the community repo doc. 

> Anyway, what's the benefit of using the same top-level type for both request and response?

That's how all the review APIs work today (including the current AdmissionReview object), we know API machinery, discovery, and clients support it well, and i didn't see a reason called out why this was different.

**Comment by caesarxuchao** [comment]
Can we trust the webhook to reset the `AdmissionReview.Spec` before sending the response? The Spec contains oldObject and Object, which is more complicated than the payloads of other review API, thus having the potential of wasting more resources if they are not cleared out.

**Comment by lavalamp** [comment]
I think @liggitt is expecting this to be a "normal" kubernetes-style API. I think we were not expecting this, as we don't expect webhook authors to be using e.g. our apiserver library or even necessarily writing their hooks in go at all, and we're optimizing for that. Additionally, I think it is a bit weird to put a mutation request in a status (status is in theory supposed to contain observables about the cluster), and it seems quite wasteful to echo the spec back in the response. Webhook calls are not declarative, they are imperative. Not a great match for Kubernetes style.

OTOH there is something to be said for being consistent. And I think that @liggitt and @deads2k are expecting to implement webhooks in regular apiservers-- for that use case it would be terribly inconvenient to have separate types.

I am not convinced that webhooks should go through the "ordinary" apiserver path, though? We could write library methods to make it easy to host just a webhook without all the other verbs that the ordinary path supports.

**Comment by liggitt** [comment]
> I think @liggitt is expecting this to be a "normal" kubernetes-style API

Yes. I'd expect the API's called by webhooks to behave like kubernetes APIs, since they are APIs defined by Kubernetes. We use api machinery and clients to generate requests and decode responses. Whether the webhook author makes use of those mechanisms or not is irrelevant.

**Comment by lavalamp** [comment]
The decoder doesn't care if the types are different, I think?

**Comment by lavalamp** [comment]
> These discussions were had at the time SubjectAccessReview was implemented. Both approaches were considered, and the current design (single type with encompassing both the input and output) was settled on. 

I wasn't part of those discussions. Sorry for relitigating. But we do have an additional need now.

> What is the justification for changing the approach for this type?

I mean, I listed pros/cons for both sides above. Which side do you think has the better arguments?

**Comment by liggitt** [comment]
Pros for a single object encompassing both the input and the result:
* Models the API call as a rest resource (post object, receive object back)
* Matches existing patterns for declarative APIs within the Kubernetes
* Allows API machinery to be used as-is to serve the webhook (while not requiring it)

At this point, I'm in favor of staying consistent and defining the object that encompasses the webhook request/response, just like the other webhook APIs

**Comment by lavalamp** [comment]
Let me clarify something, I don't actually want people building webhooks with the apiserver library by making a group name per webhook and re-registering the admission review type. I think this encourages the wrong pattern. The system is not intended to scale in the number of api types. If we want to enable that pattern, then (IMO) we need to make the aggregator understand a single GVR for webhooks and allow various servers to register one or more named objects in that group.

I think these are imperative calls and shoehorning them into a declarative interface is confusing rather than enlightening.

So, I think I do not classify your pros as pros. @erictune or @bgrant0607 might be able to convince me that the declarative style is worth it for consistency. But I don't think reuse of api machinery is ever going to be a pro in my book, as I think that is an anti pattern for this call.

There is precedent in our API for imperative calls, mostly as subresources. I think this should be modeled after those items.

**Comment by lavalamp** [comment]
Talked w/ @bgrant0607 and he pointed out that if we look at a client request path, e.g. a POST, we actually sometimes return a different object (status, which is usually but not always an error response). We are modelling the request path here, not a complete object lifecycle (admission requests don't represent long-lived stored objects), so it seems appropriate to think of it in imperative terms.

File: pkg/apis/admission/types.go
Comment (lavalamp) [inline pkg/apis/admission/types.go]: AdmissionReviewRequest, maybe?

File: pkg/apis/admission/types.go
Comment (lavalamp) [inline pkg/apis/admission/types.go]: I think we do want to send back some of the metadata to ensure that the webhook is responding to the correct response (e.g., maybe for an audit log).

File: pkg/apis/admission/types.go
Comment (lavalamp) [inline pkg/apis/admission/types.go]: I would call this just Patch.

Comment (lavalamp) [reply pkg/apis/admission/types.go]: As discussed, we need a PatchType field or something.

(we decided that this was probably a little better than extending via doing a one-of arrangement with JSONPatch, StrategicMergePatch, etc.)

File: pkg/apis/admission/types.go
Comment (lavalamp) [reply pkg/apis/admission/types.go]: It's too bad Object and OldObject are both inside the existing Spec object.

**Review by lavalamp — COMMENTED** [review]
Thoughts about the actual PR

**Comment by lavalamp** [comment]
There is one consideration which could sway me, which is that it would be a ton easier to slap a `patch` field in the status and be done with it :/

File: hack/update-all.sh
Comment (jbeda) [inline hack/update-all.sh]: Why are these commented out?

**Review by jbeda — CHANGES_REQUESTED** [review]

**Comment by jbeda** [comment]
/hold

Returning new types is a big change and the PR that makes that change is probably not the appropriate place to put this.  This is something that could impact quite a few people across the project and should follow our normal processes.

This seems to be an ideal candidate to try out the KEP process and get lots of eyes on this.  We should also think about introducing some level of style/consistency for these types of APIs.

This may be the right way to go.  But if we are going to go this direction there should be more open discussion with a proposal that calls this out explicitly.  How about you write something up (feel free to use the KEP process -- I have an updated template in flight) and we can discuss this at SIG-Architecture on Thursday?

**Comment by lavalamp** [comment]
This is an alpha api, so it affects almost nobody. We might not do what's currently in this PR. But not for that reason.

File: hack/update-all.sh
Comment (cheftako) [reply hack/update-all.sh]: Sorry. I will remove that. update-bazel hasn't worked for me since I switched to go 1.9. I just wanted to be able to run the script locally.

**Comment by lavalamp** [comment]
I filed #55729; @cheftako I think if you take a look there you will be able to predict the comments I would leave on this PR :)

**Comment by jbeda** [comment]
/hold

Please don't remove a hold like this without discussion.

There is no indication that this is an alpha API. The linked feature (that this closes) (https://github.com/kubernetes/features/issues/492) mentions beta in multiple places. This API isn't talked about anywhere in the linked proposal (https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/admission-control-webhooks.md).

Let's hold this pending the discussion in #55729.  Beyond that, we really should have design docs around new APIs -- even if they are alpha.  It would also be great to clear up the confusion around this closing out an issue that is all about bringing webhooks to *beta*.

**Comment by lavalamp** [comment]
Just to be clear, I meant the *existing* API is alpha, so we are not locked in yet. We do intend to make it beta.

**Comment by bgrant0607** [comment]
@jbeda 

Let's discuss in SIG Architecture how we want to role out the KEP process, but this is not the best way.  

With respect to OARP, 4 of 5 of the project's API approvers are already looking at this API, and the tech leads for API machinery are involved, as well.

File: pkg/kubectl/scheme/install.go
Comment (sttts) [inline pkg/kubectl/scheme/install.go]: what is kubectl doing with admission reviews?

File: staging/src/k8s.io/api/admission/v1alpha1/types.go
Comment (sttts) [inline staging/src/k8s.io/api/admission/v1alpha1/types.go]: surprising to have the same metadata again as in the request. Why?

Comment (lavalamp) [reply staging/src/k8s.io/api/admission/v1alpha1/types.go]: Explanation in #55729. But likely we are going with plan b on this one.

**Pull request closed without merging**
